scipy/signal/_peak_finding.py

```diff
@@ -336,78 +345,62 @@ def peak_prominences(x, peaks, wlen=None):
     >>> plt.plot(peaks, x[peaks], "x")
     >>> plt.vlines(x=peaks, ymin=contour_heights, ymax=x[peaks])
     >>> plt.show()
-    """
-    x = np.asarray(x)
-    peaks = np.asarray(peaks)
-    if peaks.size == 0:
-        # Handle empty peaks
-        return np.array([]), np.array([], dtype=int), np.array([], dtype=int)
 
+    Let's evaluate a second example that demonstrates several edge cases for
+    one peak at index 5.
+
+    >>> x = np.array([0, 1, 0, 3, 1, 3, 0, 4, 0])
+    >>> peaks = np.array([5])
+    >>> plt.plot(x)
+    >>> plt.plot(peaks, x[peaks], "x")
+    >>> plt.show()
+    >>> peak_prominences(x, peaks)  # -> (prominences, left_bases, right_bases)
+    (array([3.]), array([2]), array([6]))
+
+    Note how the peak at index 3 of the same height is not considered as a
+    border while searching for the left base. Instead two minima at 0 and 2
+    are found in which case the one closer to the evaluated peak is always
+    chosen. On the right side however the base must be placed at 6 because the
+    higher peak represents the right border to the evaluated area.
+
+    >>> peak_prominences(x, peaks, wlen=3.1)
+    (array([2.]), array([4]), array([6]))
+
+    Here we restricted the algorithm to a window from 3 to 7 (the length is 5
+    samples because `wlen` was rounded up to the next odd integer). Thus the
+    only two candidates in the evaluated area are the two neighbouring samples
+    and a smaller prominence is calculated.
+    """
+    # Inner function expects `x` to be C-contiguous
+    x = np.asarray(x, order='C', dtype=np.float64)
     if x.ndim != 1:
         raise ValueError('`x` must have exactly one dimension')
+
+    peaks = np.asarray(peaks)
+    try:
+        # Safely convert to C-contiguous array of type np.intp
+        peaks = peaks.astype(np.intp, order='C', casting='safe',
+                             subok=False, copy=False)
+    except TypeError:
+        if peaks.size == 0:
+            # Empty arrays default to np.float64 but are valid input
+            peaks = np.array([], dtype=np.intp)
+        else:
+            raise TypeError("Cannot safely cast `peaks` to dtype('intp')")
```

Comment:
This control-flow seems a bit complicated. I'd prefer to check for `peaks.size == 0` before the try/except block, then do the safe `astype` cast without catching the `TypeError`.

Reply:
The reasoning is that I expect `peaks` being an empty sequence to be very uncommon. Thus I felt always doing that check might be unnecessary. However, I think the readability trumps the (unnecessary) optimization in this case so I'll change it.